rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+//! # yarp
+//!
+//! Rustified version of Ruby's YARP parser.
+//!
+#![warn(
+    box_pointers,
+    clippy::all,
+    clippy::nursery,
+    clippy::pedantic,
+    future_incompatible,
+    missing_docs,
+    nonstandard_style,
+    rust_2018_idioms,
+    trivial_casts,
+    trivial_numeric_casts,
+    unreachable_pub,
+    unused_qualifications
+)]
+
+// Most of the code in this file is generated, so sometimes it generates code
+// that doesn't follow the clippy rules. We don't want to see those warnings.
+#![allow(
+    clippy::too_many_lines,
+    clippy::use_self
+)]
+
+mod bindings {
+    // In `build.rs`, we generate bindings based on the config.yml file. Here is
+    // where we pull in those bindings and make them part of our library.
+    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+}
+
+use std::ffi::c_char;
+use std::{ffi::CStr, marker::PhantomData};
+use std::mem::MaybeUninit;
+use std::ptr::NonNull;
+
+use yarp_sys::{yp_parser_t, yp_parser_init, yp_parse, yp_node_destroy, yp_node_t, yp_comment_t, yp_diagnostic_t};
+pub use self::bindings::*;
+
+/// A diagnostic message that came back from the parser.
+#[derive(Debug)]
+pub struct Diagnostic<'pr> {
+    diagnostic: NonNull<yp_diagnostic_t>,
+    marker: PhantomData<&'pr yp_diagnostic_t>
+}
+
+impl<'pr> Diagnostic<'pr> {
+    /// Returns the message associated with the diagnostic.
+    /// 
+    /// # Panics
+    ///
+    /// Panics if the message is not able to be converted into a `CStr`.
+    ///
+    #[must_use]
+    pub fn message(&self) -> String {
+        unsafe {
+            let message: *mut c_char = self.diagnostic.as_ref().message as *mut c_char;
+            CStr::from_ptr(message).to_str().unwrap().to_string()
```

Comment:
```suggestion
            let cstr = CStr::from_ptr(message);
            cstr.to_str().expect("yarp uses UTF-8 internally for comments")
```

Reply:
use a binding for the `CStr` to be defensive against future refactors which may cause that temporary to be dropped.